ml-agents/mlagents/trainers/policy/policy.py

```diff
@@ -1,12 +1,152 @@
-from abc import ABC, abstractmethod
+from abc import abstractmethod
+from typing import Dict, List, Optional
+import numpy as np
 
 from mlagents_envs.base_env import DecisionSteps
+from mlagents_envs.exception import UnityException
+
 from mlagents.trainers.action_info import ActionInfo
+from mlagents_envs.base_env import BehaviorSpec
+from mlagents.trainers.settings import TrainerSettings, NetworkSettings
 
 
-class Policy(ABC):
-    @abstractmethod
+class UnityPolicyException(UnityException):
+    """
+    Related to errors with the Trainer.
+    """
+
+    pass
+
+
+class Policy:
+    def __init__(
+        self,
+        behavior_spec: BehaviorSpec,
+        seed: int,
+        trainer_settings: TrainerSettings,
+        model_path: str,
+        load: bool = False,
+        tanh_squash: bool = False,
+        reparameterize: bool = False,
+        condition_sigma_on_obs: bool = True,
+    ):
+        self.behavior_spec = behavior_spec
+        self.trainer_settings = trainer_settings
+        self.network_settings: NetworkSettings = trainer_settings.network_settings
+        self.seed = seed
+        self.act_size = (
+            list(behavior_spec.discrete_action_branches)
+            if behavior_spec.is_action_discrete()
+            else [behavior_spec.action_size]
+        )
+        self.vec_obs_size = sum(
+            shape[0] for shape in behavior_spec.observation_shapes if len(shape) == 1
+        )
+        self.vis_obs_size = sum(
+            1 for shape in behavior_spec.observation_shapes if len(shape) == 3
+        )
+        self.model_path = model_path
+        self.initialize_path = self.trainer_settings.init_path
+        self.keep_checkpoints = self.trainer_settings.keep_checkpoints
+        self.use_continuous_act = behavior_spec.is_action_continuous()
+        self.num_branches = self.behavior_spec.action_size
```

Comment:
I would be more inclined to make these read only properties that access self.behavior_spec. I think it would simplify the __init__ method and make it clearer. I would do the same for all the public properties such as use_recurrent, h_size, etc...

Reply:
I see where you are coming from, but just want to make sure this is the way we want to go, since it would add a number of additional properties, and they all belong to the behavior_spec anyway.  Should we maybe just keep the spec itself and use that for reference? What do you think @ervteng ?